File: mira-hub/src/lib/knowledge-graph/asset-bridge.ts
/**
 * Asset → KG bridge row: make a created asset a WHOLE machine.
 *
 * WHY THIS EXISTS
 * A machine lives in two tables. `cmms_equipment` is the asset register a
 * technician sees, and the row `/api/assets/by-tag/{tag}` resolves a scanned
 * sticker against. `kg_entities` is the graph node every grounded surface
 * reads — `/api/assets/{id}/notebook` refuses any asset with no
 * `entity_type='equipment'` node that is `approval_state='verified'` AND
 * carries a non-null `uns_path`.
 *
 * Until this module, neither create door wrote both:
 *
 *   POST /api/namespace/node  → kg_entities only     → notebook ok, scan 404s
 *   POST /api/assets          → cmms_equipment only  → scan ok, notebook 404s
 *
 * CV-101 answered on a phone only because `tools/seeds/garage-cv101-kg-bridge.sql`
 * built its bridge row by hand. Every other asset failed the second call with
 * "That asset isn't in this account" — for an asset that IS in the account. An
 * acceptance test that only ever exercised CV-101 could not see it.
 *
 * WHY approval_state IS PINNED 'verified'
 * Pinned explicitly, exactly as `POST /api/namespace/node` pins it, for the
 * same two reasons:
 *
 *  1. The column default is contradictory across the two migration lineages —
 *     'proposed' at `mira-hub/db/migrations/029_kg_approval_state.sql`,
 *     'verified' at `docs/migrations/008_kg_approval_state.sql`. Relying on the
 *     default makes a machine chattable or not depending on which set won on
 *     the box you deployed to. Pinning removes the question.
 *
 *  2. It is NOT an auto-promotion. The prohibition in `.claude/CLAUDE.md`
 *     ("Do not auto-promote proposed → verified") is about MIRA *inferring* a
 *     fact and marking it reviewed without a human; the mechanically enforced
 *     Iron Rule (`scripts/kg_write_guard.py`, ADR-0017) governs
 *     `kg_relationships`. Here a human typed a machine they own into their own
 *     asset register — that act IS the approval — and no relationship is
 *     asserted. The same reasoning already licenses the pinned 'verified' in
 *     `namespace/node`, `createNotebook`, and `createAndBindNotebookTx`.
 *
 * WHAT THE SHAPE HAS TO BE (copied from the proven CV-101 seed)
 *   entity_type = 'equipment'
 *   entity_id   = cmms_equipment.id AS TEXT   ← the key the notebook route,
 *                 machine-memory, context and signal-history all match on
 *   properties  → carries 'asset_tag', the alias key that IS read
 *                 (mira-bots/shared/demo_namespace.py); 'equipment_number' is
 *                 kept alongside it because the seed wrote it, but no reader
 *                 uses that one.
 *   uns_path    → non-null, or the notebook route still refuses (see below)
 * and `cmms_equipment.uns_path` is backfilled to match, because
 * `.claude/rules/uns-compliance.md` requires every asset row to carry one.
 */

import type { PoolClient } from "pg";
import { equipmentPath, sitePath, slugify } from "@/lib/uns";
import { insertWithUniqueFallback } from "@/lib/pg-unique-retry";
import { resolveTenantParentPath } from "./cmms-sync";

/**
 * Name of the site minted for a tenant that has never run the namespace
 * wizard. Deliberately generic and renameable in the namespace builder — it
 * is a placeholder for a real hierarchy, not a claim about the plant.
 */
export const DEFAULT_SITE_NAME = "Main Site";

export type AssetBridgeInput = {
  /** cmms_equipment.id — becomes kg_entities.entity_id (as text). */
  assetId: string;
  /**
   * cmms_equipment.equipment_number — the scannable tag. Required: it is the
   * only per-tenant-unique handle available to disambiguate the natural key.
   */
  tag: string;
  description?: string | null;
  manufacturer?: string | null;
  model?: string | null;
};

export type AssetBridgeResult =
  | { ok: true; created: boolean; nodeId: string; unsPath: string }
  | { ok: false; reason: "no_uns_path" };

/**
 * Resolve the UNS path a newly created asset should live at, creating the
 * tenant's root site if it has none.
 *
 * ── THE DECISION THIS FUNCTION ENCODES ────────────────────────────────────
 * `equipmentPath(parent, ident)` returns null without a parent path, and
 * `POST /api/auth/register` seeds no site node — so a brand-new tenant (the
 * beta gate's "stranger") has NO hierarchy, and a bridge row minted for them
 * would carry a null uns_path and the notebook would still refuse it. Three
 * ways out; this function implements the first:
 *
 *   A. Auto-create a minimal root (IMPLEMENTED). Mint one `site` node at
 *      `enterprise.main_site` on the first asset and nest the machine under
 *      it. Every machine gets a genuine, renameable path; no gate is weakened;
 *      zero onboarding friction.
 *   B. Relax the notebook route's non-null uns_path check. Fastest, but a
 *      path-less machine silently breaks any later `uns_path <@` live-evidence
 *      probe and weakens `.claude/rules/uns-compliance.md`.
 *   C. Require the wizard before an asset can be created. Doctrinally cleanest
 *      and closest to train-before-deploy, but it puts a multi-step wizard in
 *      front of a technician who just wants to scan a sticker and ask.
 *
 * Switching to B or C is a change to THIS function and its caller's handling
 * of `no_uns_path` — nothing else in the bridge depends on which one is live.
 */
export async function resolveOrCreateAssetUnsPath(
  client: PoolClient,
  tenantId: string,
  tag: string,
): Promise<string | null> {
  const existing = await resolveTenantParentPath(client, tenantId);
  const parent = existing ?? (await createDefaultSite(client, tenantId));
  if (!parent) return null;
  return equipmentPath(parent, tag);
}

/**
 * Mint the tenant's root site. Idempotent via the kg_entities natural key
 * (tenant_id, entity_type, name) — the same ON CONFLICT target the onboarding
 * wizard uses, so a tenant that later runs the wizard reconciles onto this row
 * rather than duplicating it.
 */
async function createDefaultSite(client: PoolClient, tenantId: string): Promise<string | null> {
  const path = sitePath(DEFAULT_SITE_NAME);
  if (!path) return null;
  await client.query(
    `INSERT INTO kg_entities (tenant_id, entity_type, entity_id, name, properties, uns_path, approval_state)
     VALUES ($1::uuid, 'site', $2, $3, $4::jsonb, $5::ltree, 'verified')
     ON CONFLICT (tenant_id, entity_type, name) DO UPDATE
        SET uns_path = COALESCE(kg_entities.uns_path, EXCLUDED.uns_path),
            updated_at = now()`,
    [
      tenantId,
      slugify(DEFAULT_SITE_NAME),
      DEFAULT_SITE_NAME,
      JSON.stringify({ source: "asset_create_default_site" }),
      path,
    ],
  );
  return path;
}

/**
 * Create the bridge node for a just-created asset.
 *
 * Runs on the CALLER'S client so it joins the caller's transaction — an asset
 * create either produces a whole machine or nothing.
 *
 * NATURAL-KEY COLLISION. `kg_entities` is UNIQUE on
 * (tenant_id, entity_type, name) (migrations 025/026, promoted to a real
 * constraint by 064). Two machines can easily share a description
 * ("Discharge Conveyor"), and the notebooks' own backing nodes are ALSO
 * entity_type='equipment', so they compete for the same key. The tag is
 * per-tenant unique (migration 083), so on collision we disambiguate with it
 * rather than failing the create.
 */
export async function mintAssetBridgeNode(
  client: PoolClient,
  tenantId: string,
  input: AssetBridgeInput,
): Promise<AssetBridgeResult> {
  const unsPath = await resolveOrCreateAssetUnsPath(client, tenantId, input.tag);
  if (!unsPath) return { ok: false as const, reason: "no_uns_path" as const };

  // Already bridged (a retry, or an asset the CMMS→KG sync reached first).
  const existing = await client.query<{ id: string }>(
    `SELECT id::text AS id FROM kg_entities
      WHERE tenant_id = $1::uuid AND entity_type = 'equipment' AND entity_id = $2
      LIMIT 1`,
    [tenantId, input.assetId],
  );
  if (existing.rows[0]) {
    return { ok: true as const, created: false, nodeId: existing.rows[0].id, unsPath };
  }

  const base = (input.description ?? "").trim();
  const properties = JSON.stringify({
    source: "asset_create",
    asset_tag: input.tag,
    equipment_number: input.tag,
    manufacturer: input.manufacturer ?? null,
    model_number: input.model ?? null,
  });

  // `name` renders as the machine's title on the scan card, so prefer the
  // human description and fall back to a tag-qualified name only when taken.
  // The fallback MUST be savepoint-fenced: this runs inside the asset-create
  // transaction, where a 23505 aborts everything and a plain retry would die
  // with "current transaction is aborted" instead.
  const preferred = base || input.tag;
  const fallback = base ? `${base} (${input.tag})` : input.tag;
  const res = await insertWithUniqueFallback<{ id: string }>(
    client,
    `INSERT INTO kg_entities
       (tenant_id, entity_type, entity_id, name, properties, uns_path, approval_state)
     VALUES ($1::uuid, 'equipment', $2, $3, $4::jsonb, $5::ltree, 'verified')
     RETURNING id::text AS id`,
    [tenantId, input.assetId, preferred, properties, unsPath],
    [tenantId, input.assetId, fallback, properties, unsPath],
  );
  return { ok: true as const, created: true, nodeId: res.rows[0].id, unsPath };
}

/**
 * Backfill `cmms_equipment.uns_path` so the asset row carries its own path
 * (`.claude/rules/uns-compliance.md`). Only fills a NULL — never overwrites a
 * path the tenant or a backfill already chose.
 */
export async function backfillAssetUnsPath(
  client: PoolClient,
  tenantId: string,
  assetId: string,
  unsPath: string,
): Promise<void> {
  await client.query(
    `UPDATE cmms_equipment
        SET uns_path = $3::ltree, updated_at = now()
      WHERE tenant_id = $1 AND id = $2 AND uns_path IS NULL`,
    [tenantId, assetId, unsPath],
  );
}

File: mira-bots/shared/asset_bridge.py
"""Python mirror of the Hub asset bridge (#3708).

Canonical writer: ``mira-hub/src/lib/knowledge-graph/asset-bridge.ts``
(``mintAssetBridgeNode`` → ``backfillAssetUnsPath``). ``POST /api/assets`` has
minted a whole machine since #3382: the ``cmms_equipment`` row, its
``kg_entities`` equipment node, and ``cmms_equipment.uns_path``. Every Hub
consumer that needs a machine's location — the notebook, V3's scope picker,
machine memory / history / signal-history via ``resolveAssetUnsPath`` —
anchors on the ``kg_entities`` node, so a ``cmms_equipment`` row inserted
without this bridge is unplaced by construction: the register shows it and
MIRA refuses to talk about it.

Three Python writers insert ``cmms_equipment`` (bot conversations, PM
scheduling, the Atlas sync). They cannot call the TypeScript bridge, so this
module mirrors it statement for statement. Parity with the TS source is
guarded by ``mira-bots/tests/test_asset_bridge.py`` (it reads the ``.ts``
files), and ``tests/test_architecture.py`` Contract 7 refuses any Python
``cmms_equipment`` insert that bypasses ``bridge_asset``.

Works on any DB-API cursor (psycopg2 ``%s`` params). From a SQLAlchemy
connection hand over ``conn.connection.cursor()`` — same DBAPI connection,
same transaction.

Path grammar is the Hub's compact one (``enterprise.<site>.<asset>``), not the
ISA-95 form ``tools/cmms_equipment_uns_backfill.py`` writes — the bridge must
produce the same node the Hub would have.
"""

from __future__ import annotations

import json
import logging
import re
from dataclasses import dataclass
from typing import Any, Protocol

from .uns_paths import slug

logger = logging.getLogger("mira-gsd")

# Mirrors asset-bridge.ts — the parity test asserts these literals match.
DEFAULT_SITE_NAME = "Main Site"
SLUG_MAX_LEN = 64
_UNIQUE_VIOLATION = "23505"
_UUID_RE = re.compile(r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$", re.I)


def is_uuid_tenant(tenant_id: str | None) -> bool:
    """kg_entities.tenant_id is UUID; cmms_equipment.tenant_id is TEXT and still holds
    legacy slugs ('mike' — the HUB_TENANT_ID default in the compose files, and what
    pm_scheduler normalises a tenantless schedule to). A slug can never own a node."""
    return bool(tenant_id) and bool(_UUID_RE.match(tenant_id))


class DbCursor(Protocol):
    def execute(self, sql: str, params: Any = ...) -> Any: ...

    def fetchone(self) -> Any: ...


@dataclass(frozen=True)
class BridgeResult:
    ok: bool
    uns_path: str | None = None
    node_id: str | None = None
    created: bool = False
    reason: str | None = None


def hub_slug(value: str | None) -> str | None:
    """``uns.ts::slugify`` — ``uns_paths.slug`` capped at 64 chars, ``None`` when empty."""
    cleaned = slug(value or "")[:SLUG_MAX_LEN]
    return cleaned or None


def site_path(site: str) -> str | None:
    s = hub_slug(site)
    return f"enterprise.{s}" if s else None


def equipment_path(parent_path: str | None, tag: str) -> str | None:
    s = hub_slug(tag)
    return f"{parent_path}.{s}" if parent_path and s else None


def _resolve_tenant_parent_path(cur: DbCursor, tenant_id: str) -> str | None:
    """``cmms-sync.ts::resolveTenantParentPath`` — the tenant's deepest structural node."""
    cur.execute(
        """SELECT uns_path::text AS uns_path
             FROM kg_entities
            WHERE tenant_id = %s::uuid
              AND entity_type IN ('site', 'plant', 'line', 'production_line')
              AND uns_path IS NOT NULL
            ORDER BY nlevel(uns_path) DESC, updated_at DESC
            LIMIT 1""",
        (tenant_id,),
    )
    row = cur.fetchone()
    return row[0] if row and row[0] else None


def _create_default_site(cur: DbCursor, tenant_id: str) -> str | None:
    path = site_path(DEFAULT_SITE_NAME)
    if not path:
        return None
    cur.execute(
        """INSERT INTO kg_entities (tenant_id, entity_type, entity_id, name, properties, uns_path, approval_state)
           VALUES (%s::uuid, 'site', %s, %s, %s::jsonb, %s::ltree, 'verified')
           ON CONFLICT (tenant_id, entity_type, name) DO UPDATE
              SET uns_path = COALESCE(kg_entities.uns_path, EXCLUDED.uns_path),
                  updated_at = now()""",
        (
            tenant_id,
            hub_slug(DEFAULT_SITE_NAME),
            DEFAULT_SITE_NAME,
            json.dumps({"source": "asset_create_default_site"}),
            path,
        ),
    )
    return path


def resolve_or_create_asset_uns_path(cur: DbCursor, tenant_id: str, tag: str) -> str | None:
    parent = _resolve_tenant_parent_path(cur, tenant_id) or _create_default_site(cur, tenant_id)
    if not parent:
        return None
    return equipment_path(parent, tag)


def _insert_with_unique_fallback(
    cur: DbCursor, sql: str, first: tuple[Any, ...], second: tuple[Any, ...]
) -> Any:
    """``pg-unique-retry.ts::insertWithUniqueFallback`` — retry once on 23505."""
    cur.execute("SAVEPOINT sp_uniq")
    try:
        cur.execute(sql, first)
    except Exception as exc:
        cur.execute("ROLLBACK TO SAVEPOINT sp_uniq")
        if getattr(exc, "pgcode", None) != _UNIQUE_VIOLATION:
            cur.execute("RELEASE SAVEPOINT sp_uniq")
            raise
        cur.execute(sql, second)
    row = cur.fetchone()
    cur.execute("RELEASE SAVEPOINT sp_uniq")
    return row


def mint_asset_bridge_node(
    cur: DbCursor,
    tenant_id: str,
    asset_id: str,
    tag: str,
    *,
    description: str | None = None,
    manufacturer: str | None = None,
    model: str | None = None,
) -> BridgeResult:
    """``asset-bridge.ts::mintAssetBridgeNode`` — idempotent on (tenant, 'equipment', asset_id)."""
    uns_path = resolve_or_create_asset_uns_path(cur, tenant_id, tag)
    if not uns_path:
        return BridgeResult(ok=False, reason="no_uns_path")

    cur.execute(
        """SELECT id::text AS id FROM kg_entities
            WHERE tenant_id = %s::uuid AND entity_type = 'equipment' AND entity_id = %s
            LIMIT 1""",
        (tenant_id, asset_id),
    )
    existing = cur.fetchone()
    if existing:
        return BridgeResult(ok=True, uns_path=uns_path, node_id=str(existing[0]), created=False)

    base = (description or "").strip()
    properties = json.dumps(
        {
            "source": "asset_create",
            "asset_tag": tag,
            "equipment_number": tag,
            "manufacturer": manufacturer,
            "model_number": model,
        }
    )
    preferred = base or tag
    fallback = f"{base} ({tag})" if base else tag
    sql = """INSERT INTO kg_entities
               (tenant_id, entity_type, entity_id, name, properties, uns_path, approval_state)
             VALUES (%s::uuid, 'equipment', %s, %s, %s::jsonb, %s::ltree, 'verified')
             RETURNING id::text AS id"""
    row = _insert_with_unique_fallback(
        cur,
        sql,
        (tenant_id, asset_id, preferred, properties, uns_path),
        (tenant_id, asset_id, fallback, properties, uns_path),
    )
    return BridgeResult(ok=True, uns_path=uns_path, node_id=str(row[0]), created=True)


def backfill_asset_uns_path(cur: DbCursor, tenant_id: str, asset_id: str, uns_path: str) -> None:
    """``asset-bridge.ts::backfillAssetUnsPath`` — only fills a NULL column, never overwrites."""
    cur.execute(
        """UPDATE cmms_equipment
              SET uns_path = %s::ltree, updated_at = now()
            WHERE tenant_id = %s AND id = %s AND uns_path IS NULL""",
        (uns_path, tenant_id, asset_id),
    )


def bridge_asset(
    cur: DbCursor,
    tenant_id: str,
    asset_id: str,
    tag: str,
    *,
    description: str | None = None,
    manufacturer: str | None = None,
    model: str | None = None,
) -> BridgeResult:
    """The whole bridge for one freshly inserted ``cmms_equipment`` row.

    Runs inside a SAVEPOINT so a bridge failure rolls back to the caller's
    state and never aborts the enclosing transaction (a bot work order must
    still land when the KG write fails). A failure is logged, exactly as the
    Hub route warns, and reported in the result — never raised.

    A legacy (non-UUID) tenant is refused BEFORE any SQL: the ``::uuid`` cast
    would fail inside the savepoint and read like a transient error, when it
    is a structural one — that machine stays unplaced until its tenant
    migrates (Codex F1 on #3715). The repair path for rows already written
    this way is tools/cmms_equipment_uns_backfill.py, which reports them.
    """
    if not is_uuid_tenant(tenant_id):
        logger.warning(
            "asset_bridge: asset %s under legacy tenant %r cannot be bridged — "
            "kg_entities.tenant_id is UUID-only; the row stays UNPLACED until the tenant "
            "migrates (#3708)",
            asset_id,
            tenant_id,
        )
        return BridgeResult(ok=False, reason="legacy_tenant")
    cur.execute("SAVEPOINT asset_bridge")
    try:
        result = mint_asset_bridge_node(
            cur,
            tenant_id,
            asset_id,
            tag,
            description=description,
            manufacturer=manufacturer,
            model=model,
        )
        if result.ok and result.uns_path:
            backfill_asset_uns_path(cur, tenant_id, asset_id, result.uns_path)
        cur.execute("RELEASE SAVEPOINT asset_bridge")
    except Exception as exc:
        cur.execute("ROLLBACK TO SAVEPOINT asset_bridge")
        cur.execute("RELEASE SAVEPOINT asset_bridge")
        logger.warning(
            "asset_bridge: asset %s bridge failed (%s: %s)", asset_id, type(exc).__name__, exc
        )
        return BridgeResult(ok=False, reason=f"error:{type(exc).__name__}")
    if not result.ok:
        logger.warning(
            "asset_bridge: asset %s created without a UNS path (%s) — notebook will refuse it",
            asset_id,
            result.reason,
        )
    return result
